binary_search_tree/tree.py

```diff
@@ -17,41 +18,103 @@ def __init__(self):
     # Time Complexity: 
     # Space Complexity: 
     def add(self, key, value = None):
-        pass
+        new_node = TreeNode(key, value)
+        if self.root is None:
+            self.root = new_node
+            return self.root
+        return self.add_helper(self.root, new_node)
+
+    def add_helper(self, curr, new_node):
+        if curr is None:
+            curr = new_node
+        elif new_node.key < curr.key:
+            curr.left = self.add_helper(curr.left, new_node)
+        else:
+            curr.right = self.add_helper(curr.right, new_node)
+        return curr
```

Comment:
This implementation of add_helper works, but definitely does a bit of extra work in setting nodes. Essentially, every node on the path to where the new node ends up will have either its left or right node "updated", though most of these will be updated to the exact same node. The only one that gets a meaningful update is whenever the bottom of the tree is reached and gets its left/right node updated to the new node. These updates happen in lines 31 and 32.

This probably will not be too bad of a performance hit, but there is a way that you can do this and just update only the node you need.

But this overall fine enough!